### PR TITLE
allocate-ipip-addr: retry loop around IPAM AutoAssign()

### PR DESCRIPTION
**Pull request by ti-mo**
Has a random sub-second delay to prevent the (unlikely) perpetuation
of the race condition throughout the retry loop.

## Description

Workaround for the race condition outlined in https://github.com/projectcalico/node/issues/9. All relevant details are mentioned there.

I don't have sufficient knowledge of the project to write a meaningful unit test. The 'why' is explained in the code.

## Todos
- [ ] Tests
- [ ] Documentation
- [x] Release note

## Release Note

```release-note
Internally retry address allocation from IP Pools with IPIP enabled (impacts pre-3.0 releases)
```

@caseydavenport

### CODE REVIEW

**Comment by ti-mo** [comment]
@caseydavenport Since this is likely only useful for 2.6, should it be backported to https://github.com/projectcalico/calico `release-v2.6`?

File: pkg/allocateipip/allocate_ipip_addr.go [hunk 1]
```diff
@@ -159,9 +160,20 @@ func assignHostTunnelAddr(ctx context.Context, c client.Interface, nodename stri
...
+	// Retry loop around AutoAssign to recover from a distributed race in case
+	// all nodes in a large cluster respond to a change in /pools simultaneously.
+	for a := 3; a > 0; a-- {
+		var err error
+		ipv4Addrs, _, err = c.IPAM().AutoAssign(ctx, args)
+
+		if err != nil {
+			log.WithError(err).Error("Unable to autoassign an address for IPIP:")
+			time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond)
+		} else {
```
Comment (caseydavenport) [inline pkg/allocateipip/allocate_ipip_addr.go]: minor nit: I don't think we need the else block around the break here, so I'd be in favor of removing it to make reading easier.

**Comment by caseydavenport** [comment]
> Since this is likely only useful for 2.6 or earlier, should it be backported to https://github.com/projectcalico/calico release-v2.6?

@ti-mo yep, I think it should be. Unfortunately the `calico/node` code moved repos between v3.1 and (upcoming) v3.2, so you'll need to open a PR over in the Calico repo.

As for this PR, I'm not sure it's needed? I believe that in v3 the IPAM code does the necessary retries.

**Comment by caseydavenport** [comment]
Example retry in v3: https://github.com/projectcalico/libcalico-go/blob/master/lib/ipam/ipam.go#L281-L285

In v2.6, we don't retry: https://github.com/projectcalico/libcalico-go/blob/release-v2.6/lib/client/ipam.go#L190-L223

File: pkg/allocateipip/allocate_ipip_addr.go (hunk 1 above)
Comment (ti-mo) [reply pkg/allocateipip/allocate_ipip_addr.go]: But of course, will fix in the new PR.

**Comment by ti-mo** [comment]
Closing in favor of https://github.com/projectcalico/calico/pull/2038.

**Pull request closed without merging**